### PR TITLE
Provide option to update outputs of a logger.

### PR DESCRIPTION
**Pull request by shobhit85**
Updating an existing output of a logger might be needed to update
fields logged in a log message by the outputs of specific classes.

@krobertson @olegshaldybin

### CODE REVIEW

File: logger.go [hunk 1]
```diff
@@ -145,6 +146,46 @@ func (logger *Logger) AddOutput(uri string, classes ...LogClass) error {
 	return nil
 }
 
+// UpdateOutput updates an existing output matching for an URI scheme
+// and exact classes match.
+func (logger *Logger) UpdateOutput(uri string, classes ...LogClass) error {
+
+	// Parse the uri to search for existing Output.
```
Comment (alextoombs) [inline logger.go]: nit: comments are a little excessive.

File: logger.go [hunk 2]
```diff
@@ -145,6 +146,46 @@ func (logger *Logger) AddOutput(uri string, classes ...LogClass) error {
...
+			lo, err := createOutputWrapper(uri, classes)
+			if err != nil {
+				return err
+			}
+			if lo == nil {
+				return nil
+			}
+
+			// Lock our outputs
+			logger.outputMutex.Lock()
```
Comment (alextoombs) [inline logger.go]: Will this deadlock if you have multiple `logger.outputs` that match? You've deferred the unlock here. Maybe explicitly unlock after overwriting the value?

**Comment by alextoombs** [comment]
Needs @jpoler's opinion!

File: logger.go [hunk 3]
```diff
@@ -145,6 +146,44 @@ func (logger *Logger) AddOutput(uri string, classes ...LogClass) error {
...
+		return err
+	}
+
+	class := NONE
+	for _, c := range classes {
+		class |= c
+	}
+
+	for i, o := range logger.outputs {
+		if class&o.Class == class && u.Scheme == o.OutputWrapper.URL.Scheme {
```
Comment (olegshaldybin) [inline logger.go]: why not `o.Class == class`?

File: logger.go [hunk 4]
```diff
@@ -145,6 +146,44 @@ func (logger *Logger) AddOutput(uri string, classes ...LogClass) error {
...
+	u, err := url.Parse(uri)
+	if err != nil {
+		return err
+	}
+
+	class := NONE
+	for _, c := range classes {
+		class |= c
+	}
+
```
Comment (olegshaldybin) [inline logger.go]: It seems the whole for loop needs to be locked, as it writes back to the index where the output has been found, and that could be changed by a concurrent insertion/deletion.

File: logger.go (hunk 3 above)
Comment (olegshaldybin) [inline logger.go]: e.g. function doc says it does update on an exact match, but here more classes could be present in the output than passed via arguments.

File: logger.go [hunk 5]
```diff
@@ -145,6 +146,44 @@ func (logger *Logger) AddOutput(uri string, classes ...LogClass) error {
...
+		if class&o.Class == class && u.Scheme == o.OutputWrapper.URL.Scheme {
+			// Found a match
+
+			// generate it with new uri.
+			lo, err := createOutputWrapper(uri, classes)
+			if err != nil {
+				return err
+			}
+			if lo == nil {
+				return nil
```
Comment (olegshaldybin) [inline logger.go]: What if there is another match?

File: logger.go (hunk 3 above)
Comment (shobhit85) [inline logger.go]: Probably should remove the word "exact". Classes are ORed for a logger in on bit sequence so have to check using AND operation.

File: logger.go (hunk 4 above)
Comment (krobertson) [inline logger.go]: Agree

Comment (krobertson) [inline logger.go]: Looks like it was done already? Was looking at something else and thought it hadn't.

**Comment by krobertson** [comment]
LGTM, once verified on the test run.

**Comment by olegshaldybin** [comment]
LGTM

**Comment by shobhit85** [comment]
Added one unit test and verified it on test run.

**Pull request merged**